server/src/main/java/org/opensearch/indices/IndicesRequestCache.java

```diff
@@ -371,40 +412,300 @@ public int hashCode() {
+                    // decrement the stale key count
+                    staleKeysCount.decrementAndGet();
+                    // Remove the key if the new value is zero by returning null; otherwise, update with the new value.
+                    return newValue == 0 ? null : newValue;
+                });
+            }
+        }
+
+        /**
+         * Updates the count of stale keys in the cache.
+         * This method is called when a CleanupKey is added to the keysToClean set.
+         *
+         * It increments the staleKeysCount by the count of the CleanupKey in the cleanupKeyToCountMap.
+         * If the CleanupKey's readerCacheKeyId is null or the CleanupKey's entity is not open, it increments the staleKeysCount
+         * by the total count of keys associated with the CleanupKey's ShardId in the cleanupKeyToCountMap and removes the ShardId from the map.
+         *
+         * @param cleanupKey the CleanupKey that has been marked for cleanup
+         */
+        private void updateStaleKeysCount(CleanupKey cleanupKey) {
+            if (stalenessThreshold == 0.0) {
+                return;
+            }
+            IndexShard indexShard = (IndexShard) cleanupKey.entity.getCacheIdentity();
+            if (indexShard == null) {
+                logger.warn("IndexShard is null for CleanupKey: {}", cleanupKey.readerCacheKeyId);
+                return;
+            }
+            ShardId shardId = indexShard.shardId();
+
+            // Using computeIfPresent to atomically operate on the countMap for a given shardId
+            cleanupKeyToCountMap.computeIfPresent(shardId, (key, countMap) -> {
+                if (cleanupKey.readerCacheKeyId == null) {
+                    // Aggregate and add to staleKeysCount atomically if readerCacheKeyId is null
+                    int totalSum = countMap.values().stream().mapToInt(Integer::intValue).sum();
+                    staleKeysCount.addAndGet(totalSum);
+                    // Return null to automatically remove the mapping for shardId
+                    return null;
+                } else {
+                    // Update staleKeysCount based on specific readerCacheKeyId, then remove it from the countMap
+                    countMap.computeIfPresent(cleanupKey.readerCacheKeyId, (k, v) -> {
+                        staleKeysCount.addAndGet(v);
+                        // Return null to remove the key after updating staleKeysCount
+                        return null;
+                    });
+
+                    // Check if countMap is empty after removal to decide if we need to remove the shardId entry
+                    if (countMap.isEmpty()) {
+                        return null; // Returning null removes the entry for shardId
+                    }
+                }
+                return countMap; // Return the modified countMap to keep the mapping
+            });
+        }
+
+        AtomicInteger getStaleKeysCountForTesting() {
```

Comment:
No need to add "forTesting". Keeping it package private itself implies that?

Reply:
I thought this would be very clear to the caller, but i dont see any other implementation in the repo, i'll just add a comment